File: python/141-150/Sort List.py
class Solution:
    def merge(self,l1,l2):
        if l1==None:
            return l2
        elif l2==None:
            return l1
        p1,p2=l1,l2
        result = ListNode(0)
        cur = result
        while p1!=None and p2!=None:
            if p1.val<p2.val:
                cur.next=p1
                p1=p1.next
            else:
                cur.next=p2
                p2=p2.next
            cur=cur.next
        if p1!=None:
            cur.next=p1
        if p2!=None:
            cur.next=p2
        return result.next
    def mergesort(self,head,size):
        if size<2:
            return head
        l1 = head
        p=head
        for i in range(size//2-1):
            p=p.next
        l2=p.next
        p.next=None
        r1 = self.mergesort(l1,size//2)
        r2 = self.mergesort(l2,size-size//2)
        return self.merge(r1,r2)
    # @param head, a ListNode
    # @return a ListNode
    def sortList(self, head):
        size=0
        p = head
        while p!=None:
            size+=1
            p=p.next
        if size <2:
            return head
        return self.mergesort(head,size)
        

File: Cpp/141-150/sortList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
 
 // quick sort
class Solution1 {
public:
    ListNode* sortList(ListNode* head) {
        
    }
};

// merge sort
class Solution2 {
public:
    ListNode* merge(ListNode* l1, ListNode* l2){
     if(l1==NULL)
      return l2;
     if(l2 == NULL)
      return l1;
     ListNode* head = new ListNode(0);
     ListNode* temp = head;
     while(l1!=NULL && l2!=NULL){
      if(l1->val>l2->val){
       temp->next = l2;
       temp = temp->next;
       l2 = l2->next;
       temp->next = NULL;
     }
     else{
       temp->next = l1;
       temp = temp->next;
       l1 = l1->next;
       temp->next = NULL;
     }
    }
    if(l1==NULL)
     temp->next = l2;
    else temp->next=l1;
    return head->next;
    }
    ListNode* sortList(ListNode* head) {
        return mergeSort(head);
    }
    ListNode* mergeSort(ListNode* head){
      if(head==NULL || head->next==NULL)
       return head;
      ListNode * slow=head,*fast = head->next;
      while(fast!=NULL){
        fast = fast->next;
        if(fast==NULL)
            break;
        fast = fast->next;
        slow= slow->next;
      }
      fast = slow->next;
      slow->next = NULL;
      ListNode* l1 = mergeSort(head);
      ListNode* l2 = mergeSort(fast);
      return merge(l1,l2);
      }
};

//insert sort
class Solution3 {
public:
    ListNode* sortList(ListNode* head) {
        
    }
};

//radix sort
class Solution4 {
public:
    ListNode* sortList(ListNode* head) {
        
    }
};
